### ISSUE TITLE
Headerbars (integrated app bar and window decorations)

### ISSUE THREAD

**Issue by iBelieve** [lirios/fluid#29]
![](https://material-design.storage.googleapis.com/publish/material_v_8/material_ext_publish/0Bx4BSt6jniD7cHV2LWpmTnRUbm8/layout_structure_system_status5.png)

https://material.google.com/layout/structure.html#structure-system-bars

**Comment by SunRain** [comment]
Hi, if integrated app bar and window decorations, should we remove windows manager(like openbox,etc) in Qt quick window component,  and extend a sub qml window, implement our own "move-window, change-window-size) functions? @iBeliever @plfiorini

**Comment by SunRain** [comment]
just like what @iBeliever did in qml-material

**Comment by SunRain** [comment]
I mean WindowFlags like Qt::CustomizeWindowHint, Qt::FramelessWindowHint, etc(Can't remeber clearly)

**Comment by plfiorini** [comment]
@SunRain On the qmlOS desktop we can handle that from the platform theme plugin with a Wayland protocol that would tell the compositor weather it should decorate the window or whether the window does it by itself. On other desktops we should expect that the window manager will decorate the window hence the headerbar somehow should turn into a normal toolbar.

**Comment by SunRain** [comment]
Sorry, don't know about wayland related things, you mean this would handled by greenisland(or where is the  platform theme plugin) ? @plfiorini

**Comment by SunRain** [comment]
The workspace/platformtheme part in https://github.com/qmlos/workspace  ?

**Comment by plfiorini** [comment]
@SunRain the platform theme plugin in the workspace repository (https://github.com/qmlos/workspace) could set an env variable to disable Qt client side dialogs when it detects the application wants to decorate itself using the headerbar. The headerbar also need to communicate back to the application to move the window when dragging the bar.

**Comment by plfiorini** [comment]
@SunRain we then need to set a window property on windows that use the header bar so that the compositor knows which windows decorate themselves and which one needs a server-side decoration

**Comment by SunRain** [comment]
Hi, is it possbile to only customize the Control.ToolBar to implement this, like this sample http://paste.ubuntu.com/23128509/
I only set window flag to Qt.FramelessWindowHint, and handle mouse event in it. The problem is I need to add app icon to system(OS) task manger bar when app hide, and show app when clicking  OS task manger bar. I think these problems chould be solved by implementing from some dbus event(or some common OS libraries)?

**Comment by SunRain** [comment]
Ignore this.

**Comment by plfiorini** [comment]
@SunRain Yeah Qt.FramelessWindowHint should work on all desktops, and X11 as well.

We don't need to use platform specific code for minimize, maximize and close, we can do it with the Qt API.

However what we need is Qt hooks to trigger window move and resize from the client side QML decoration.

QPlatformWindow has a startSystemResize() method to trigger platform specific code for resize, and once this patch is integration it will have a startSystemMove() as well: https://codereview.qt-project.org/#/c/150399/

**Comment by plfiorini** [comment]
Moving windows from the decoration will be possible with Qt 5.11 (on X11 and Wayland at least).

https://imgur.com/avidazd

**Comment by damianatorrpm** [comment]
Do you have some working example code? I would like to test working qml csd.

**Comment by plfiorini** [comment]
Not yet because I don't have a Qt 5.11 build right now.

**Comment by milohr** [comment]
is this yet ready for testing? looks great!

**Comment by ToxicDragon** [comment]
I would really love to see CSD as part of Liri's apps. I guess having it as part of fluid would define these UX as a more generic approach. As for a technical proof of concept see https://github.com/johanhelsing/qt-csd-demo

It seems that the needed functionality is about to be included in Qt 5.15 (https://bugreports.qt.io/browse/QTBUG-73011)

**Comment by ToxicDragon** [comment]
Qt 5.15 alpha was just released: https://www.qt.io/blog/qt-5.15-alpha-released

The qt-csd-demo mentioned in my last post was already updated aswell.

**Comment by plfiorini** [comment]
There is still something missing in Qt.
As part of CSD we want to draw drop shadows as a margin around the actual window contents, but there is no API to set this.
For now we can implement client-side decorations in Fluid without drop shadows and wait for Qt 6, because Qt 5.15 is the last 5.x iteration and we can't add a new feature now.

**Comment by ToxicDragon** [comment]
Sounds like a plan! Rome wasn't built in a single day aswell :) if I can help in any means but coding, please let me know

**Comment by plfiorini** [comment]
Obsolete issue with Fluid 2 being in development
